src/animation/mod.rs

```diff
@@ -68,18 +68,21 @@ impl PlayingAnimation2D {
         }
     }
 
-    /// Check if the animation has just finished.
+    /// Check if the animation has just finished, based on its repetition behavior and the number of times it has repeated.
     ///
+    /// Note: An animation with `RepeatAnimation::Forever` will never finish.
     /// Note: This needs to be called in the `bevy_app::main_schedule::Update` schedule.
     #[inline]
     pub fn just_finished(&self) -> bool {
-        self.completions_this_update > 0
+        self.finished() && self.just_finished_cycle()
     }
 
-    /// How many completions the animation had this update.
+    /// Check if the animation has just finished a cycle.
+    ///
+    /// Note: This needs to be called in the `bevy_app::main_schedule::Update` schedule.
```

Comment:
From what I understand, the `self.completions_this_update` gets reset every `Update` loop, so if we check call this function before the internal Animation Set then we will have a frame delay? It would still work the frame after, but just something to be aware. Actually, couldn't this cause possible issues when no specific ordering is given? So when the user calls this function and the ECS puts the user function before the internal Animation set, and then next frame it's called after the animation set, then you would miss this?

Regardless, I would probably add a note that the user should be aware to check this after/before the animation set (unless I missed something).

Reply:
I was under the impression that the scheduling is "fixed" in the sense that it does not reorder stuff between frames. Otherwise just_pressed for Input would be subject to losing information as well.
A note that it should ideally be called after the animation set makes sense, though.

Reply:
Okay so I was under the same impression, but while testing something else I noticed that the scheduling is actually not fixed (in case you care: https://github.com/PraxTube/rantoni/commit/b9dcb2625776e096c55cf34fb999340a100922aa).

Reply:
Thanks. Could you give me some pointers on what to look for? The move_player system sometimes runs before and sometimes after some other (which?) system?

Reply:
Ah yeah sorry, didn't thing you would actually care haha. Yes, the `movement::move_player` and the `auto_input::move_player` are not running in a fixed way. I actually asked on the bevy discord server (which is a great way to learn more about the bevy internals) and I got this answer

> It can be different each time the schedule is built, which happens at least once at App initialisation (when you run the program), and I think maybe can also happen later during runtime in the right circumstances but I'm not certain on that. I'm too tired, that's wrong lol 

 >If they aren't explicitly ordered and don't have overlapping access, then they can be run concurrently in different threads so any ordering is possible each frame, including interleaving. At least that's true for the multithreaded executor, idk how the single-thread executor works.

> If they have overlapping access (e.g. both mutably access the same resource, or have overlapping mutable queries) then I think it's what I said before about it being decided when the schedule is built at App initialisation, which means it can be different each time you run the program but stays stable during runtime unless certain conditions I'm unsure of are met.

Reply:
Yeah I am also really unsure. System schedules are one of the things that always gets me. A lot of bugs happen because of this. I wonder if you could visualize this in a graph or something similar to get an easier understanding of the whole system schedule in your game. I used a simple directed graph to visualize branching dialogue in one of my games which was super helpful, though I could imagine that system schedules might be a bit more independent and thus not great for a graph.